### PR TITLE
8282319: java.util.Locale method to stream available Locales

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
This PR adds a new method to java.util.Locale which returns Stream\<Locale>

The contents of the Stream are composed of the same underlying elements as Locale.getAvailableLocales().

This method allows streaming of the Locale array without creating a defensive copy.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8301293](https://bugs.openjdk.org/browse/JDK-8301293) to be approved

### Issues
 * [JDK-8282319](https://bugs.openjdk.org/browse/JDK-8282319): java.util.Locale method to stream available Locales
 * [JDK-8301293](https://bugs.openjdk.org/browse/JDK-8301293): java.util.Locale method to stream available Locales (**CSR**)

### Reviewers
 * [Sergey Tsypanov](https://openjdk.org/census#stsypanov) (@stsypanov - Author) ⚠️ Review applies to [6df5c168](https://git.openjdk.org/jdk/pull/12269/files/6df5c16845ce7ef1b57fdb25f2422db7bdd60e84)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [b2c08e1c](https://git.openjdk.org/jdk/pull/12269/files/b2c08e1cff6563030e3aa605c028f47ef4fbc177)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12269/head:pull/12269` \
`$ git checkout pull/12269`

Update a local copy of the PR: \
`$ git checkout pull/12269` \
`$ git pull https://git.openjdk.org/jdk pull/12269/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12269`

View PR using the GUI difftool: \
`$ git pr show -t 12269`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12269.diff">https://git.openjdk.org/jdk/pull/12269.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java [hunk 1]
```diff
@@ -138,15 +140,26 @@ private static class AllAvailableLocales {
...
 
+    /**
+     * Returns a stream of available locales for all the provider classes.
+     * This stream is constructed from all the locales
+     * that are provided by each provider, including the JRE.
+     *
+     * @return a stream of the available locales for all provider classes
+     */
+    public static Stream<Locale> streamAvailableLocales() {
+        return AllAvailableLocales.AvailableLocalesSet.stream();
```
Comment (stsypanov) [inline src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java:160]: I think we could use `Stream.of(AllAvailableLocales.availableLocalesArray)` here and drop set field.

Comment (justin-curtis-lu) [reply src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: Yes you're right, I had misinterpreted the original bug description. Set is not required, and in fact the array should be streamed directly as you said.

**Review by stsypanov — APPROVED** [review]

File: src/java.base/share/classes/java/util/Locale.java [hunk 2]
```diff
@@ -1210,6 +1211,23 @@ public static Locale[] getAvailableLocales() {
...
+    /**
+     * {@return A stream of installed locales.}
+     *
+     * The returned stream represents the union of locales supported
+     * by the Java runtime environment and by installed
+     * {@link java.util.spi.LocaleServiceProvider LocaleServiceProvider}
+     * implementations. At a minimum, the returned stream must contain a
+     * {@code Locale} instance equal to {@link Locale#ROOT Locale.ROOT} and
+     * a {@code Locale} instance equal to {@link Locale#US Locale.US}.
+     * @implNote Unlike {@code getAvailableLocales()}, this method does
```
Comment (naotoj) [inline src/java.base/share/classes/java/util/Locale.java:1223]: I'd insert a blank line between the method description paragraphs and the following javadoc tags.

File: src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java [hunk 3]
```diff
@@ -147,6 +147,17 @@ private AllAvailableLocales() {
         }
     }
 
+    /**
+     * Returns a stream of available locales for all the provider classes.
+     * This stream is constructed from all the locales
+     * that are provided by each provider, including the JRE.
+     *
+     * @return a stream of the available locales for all provider classes
```
Comment (naotoj) [inline src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java:155]: Could use {@return } as well.

File: test/jdk/java/util/Locale/StreamAvailableLocales.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+/*
+ * @test
+ * @summary Test the implementation
+ * of Locale.streamAvailableLocales()
```
Comment (naotoj) [inline test/jdk/java/util/Locale/StreamAvailableLocales.java:26]: `availableLocales()`? There are other locations using the name too.

**Review by naotoj — COMMENTED** [review]
Looks good. Some minor comments follow.

**Review by naotoj — APPROVED** [review]

File: src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java [hunk 5]
```diff
@@ -148,11 +148,11 @@ private AllAvailableLocales() {
     }
 
     /**
+     * {@return a stream of the available locales for all provider classes}
+     *
      * Returns a stream of available locales for all the provider classes.
      * This stream is constructed from all the locales
      * that are provided by each provider, including the JRE.
-     *
-     * @return a stream of the available locales for all provider classes
```
Comment (LanceAndersen) [inline src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: Seems weird to have the @return where you moved it to

Comment (justin-curtis-lu) [reply src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: From what I saw in other methods, `@return ...` goes on the bottom (like the line I removed), but the javadoc 16 tag `{@return ... }` goes on the top.

File: src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java [hunk 6]
```diff
@@ -147,6 +147,17 @@ private AllAvailableLocales() {
         }
     }
 
+    /**
+     * {@return a stream of the available locales for all provider classes}
+     *
+     * Returns a stream of available locales for all the provider classes.
```
Comment (naotoj) [inline src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java:153]: This line can be removed

File: src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java (hunk 5 above)
Comment (LanceAndersen) [reply src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: OK but then we should be consistent and update all of the methods to use the new tag or  revert the change for consistency

Comment (justin-curtis-lu) [reply src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: Makes sense, will look at both Locale and LocaleServiceProviderPool and update the methods that are applicable for {@return ..}. Thanks Lance

Comment (RogerRiggs) [reply src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: That's getting a bit far afield of the original issue; a separate issue might be a better way to address those cases. $.02

Comment (justin-curtis-lu) [reply src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: Hi Lance and Roger, I have switched the tag {@return to @return for consistency. And like you proposed Roger, I will open a separate issue to address converting the tags.

**Review by LanceAndersen — APPROVED** [review]

File: src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java (hunk 5 above)
Comment (LanceAndersen) [reply src/java.base/share/classes/sun/util/locale/provider/LocaleServiceProviderPool.java]: > Hi Lance and Roger, I have switched the tag {@return to @return for consistency. And like you proposed Roger, I will open a separate issue to address converting the tags.

Thank you Justin

**Review by RogerRiggs — APPROVED** [review]

**Pull request closed without merging**